### DISCUSSION TITLE
RFC: Branching/Review workflow in TinaCMS

### DISCUSSION THREAD

**Discussion by jamespohalloran** [tinacms/tinacms#3509] [Experimental]
We recently worked on a "Hackathon" project where we (re)introduced some branching functionality into Tina. 

- When users save (on their main branch), they are asked if they want to create a preview environment (a branch) instead of pushing to production.
- Once a user creates a branch, a PR is auto-created with a Vercel preview URL (or other hosting providers in the future). 
- Users can grab to the preview URL from Tina's UI.

### Demo
https://www.loom.com/share/536b98e4f888492ea833e8a9200dc13b

### Added benefits for "Non-react" sites
Also we've also noted that "the ability to preview non-React" sites is a big feature-gap from Forestry to Tina. I wonder if this might fill that gap by integrating Netlify/Vercel's preview deployments.

This is still early, but we'd love to know if this type of workflow would be useful to your team. We try and be careful about exposing too much of the "Git" complexities on editors, but we've also heard from teams who don't want to work from a single branch. 
Let us know your thoughts!

**Comment by Perni1984** [comment]
As long as this stays an opt-in feature - I am totally sold on this! 

Personally I see the branching workflow and the preview feature as two separate features. I defenitely can imagine having the `building`/`preview` status even on the main branch (without any branching functionality) - in this case I would imagine that there is no dropdown showing the current branch, but only the `building`/`preview` button.

**Comment by mdancho84** [comment]
Definitely important for me to have the Preview functionality in Jekyll and Hugo sites. That was a really cool feature of Forestry that was a massive productivity booster for me personally (don't know about everyone else).

**Comment by jtfairbank** [comment]
Hey @jamespohalloran!  I'm a maintainer for [Distribute Aid's landing site](https://github.com/distributeaid/distributeaid.org), and planning our migration from Forestry to Tina.  This sounds really helpful for content editor-driven workflows.  It'd be helpful to have a way for editors to choose to work off of existing branches as well:

1. they could collaborate with eachother (ex: one editor makes a branch, another contributes to it)
2. they could collaborate with devs (ex: dev submits a new PR, an editor can select that branch and fill in content before merging into main)

Our current **stack** for context: Forestry (upgrading to TIna), Gatsby, React, Tailwind, Netlify

Our current **collab flow** is:

1. Dev makes a branch + work-in-progress PR.
3. This triggers a Netlify preview that will update as commits are pushed to the branch.
4. Dev adds the branch in Forestry.
5. Devs & content editors can now both contribute at the same time to build out a new section of the site.

**Pain points** in the current collab flow:

* Awkward to have to add a branch to Forestry manually.  Would like to automatically create one for each PR, just like Netlfiy automatically creates a preview when we make a new PR.
* Editors can't make their own branch from Forestry.

**Comment by bradystroud** [comment]
cc @adamcogan @wicksipedia @tkapa
Hey @jamespohalloran 

The branching feature is important to anyone who wants to see the changes attributed to the person who made them.

E..g see the green contributions on my profile 
https://github.com/bradystroud
Also shown on 
https://www.ssw.com.au/people/brady-stroud

⚠️ There is a complex workaround which decreases the UX - Users need to make a new branch with TinaCMS, make their changes then move to GitHub and create a PR to merge changes into main. Creating the PR puts their name on the change.

🤕 There is also a loss of data experince for the user - If users try make changes on a protected branch, they can use the CMS to edit but get an ❌ error when they try save... Losing the changes they made.

💡 Suggestion # 1 - Users should be pushed to create a branch before making edits.
💡 Suggestion # 2 - Allow users to create PRs from TinaCMS (and be attributed with the changes)
